File: app/src/main/java/com/mandarin/bcu/androidutil/unit/adapters/DynamicExplanation.java
package com.mandarin.bcu.androidutil.unit.adapters;

import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.mandarin.bcu.R;
import com.mandarin.bcu.androidutil.StaticStore;

import java.util.Objects;

import common.system.MultiLangCont;

public class DynamicExplanation extends Fragment {
    View view;

    public static DynamicExplanation newInstance(int val, int id,String [] titles) {
        DynamicExplanation explanation = new DynamicExplanation();
        Bundle bundle = new Bundle();
        bundle.putInt("Number",val);
        bundle.putInt("ID",id);
        bundle.putStringArray("Title",titles);
        explanation.setArguments(bundle);
        return explanation;
    }

    int val;
    int id;
    String[] titles;
    TextView unitname;
    TextView[] explains = new TextView[3];

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle bundle) {
        view = inflater.inflate(R.layout.unit_info_tab,container,false);
        val = getArguments().getInt("Number",0);
        id = getArguments().getInt("ID",0);
        titles = getArguments().getStringArray("Title");
        String[] explanation = MultiLangCont.FEXP.getCont(StaticStore.units.get(id).forms[val]);
        if(explanation == null) {
            explanation = new String[]{"","",""};
        }

        unitname = view.findViewById(R.id.unitexname);
        int[] lineid = {R.id.unitex0,R.id.unitex1,R.id.unitex2};
        for(int i=0;i<lineid.length;i++)
            explains[i] = view.findViewById(lineid[i]);

        explains[2].setPadding(0,0,0,StaticStore.dptopx(24f, Objects.requireNonNull(getActivity())));

        String name = MultiLangCont.FNAME.getCont(StaticStore.units.get(id).forms[val]);

        if(name == null)
            name = "";

        unitname.setText(name);

        for(int i=0;i < explains.length;i++) {
            if(i >= explanation.length) {
                explains[i].setText("");
            } else {
                if(explanation[i] != null)
                    explains[i].setText(explanation[i]);
            }
        }

        return view;
    }
}

File: app/src/main/java/com/mandarin/bcu/androidutil/unit/adapters/DynamicExplanation.kt
package com.mandarin.bcu.androidutil.unit.adapters

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.fragment.app.Fragment
import com.mandarin.bcu.R
import com.mandarin.bcu.androidutil.StaticStore
import common.io.json.JsonEncoder
import common.pack.Identifier
import common.util.lang.MultiLangCont
import common.util.unit.Unit

class DynamicExplanation : Fragment() {
    companion object {
        @JvmStatic
        fun newInstance(`val`: Int, data: Identifier<Unit>, titles: Array<String?>?): DynamicExplanation {
            val explanation = DynamicExplanation()
            val bundle = Bundle()

            bundle.putInt("Number", `val`)
            bundle.putStringArray("Title", titles)
            bundle.putString("Data", JsonEncoder.encode(data).toString())

            explanation.arguments = bundle

            return explanation
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, bundle: Bundle?): View? {
        val view = inflater.inflate(R.layout.unit_info_tab, container, false)

        val arg = arguments ?: return view

        val data = StaticStore.transformIdentifier<Unit>(arg.getString("Data")) ?: return view

        val fid = arg.getInt("Number", 0)

        val u = data.get() ?: return view

        var explanation = if(u.id.pack == Identifier.DEF) {
            MultiLangCont.getStatic().FEXP.getCont(u.forms[fid])
        } else {
            u.forms[fid].explanation.split("<br>").toTypedArray()
        }

        println(explanation.contentToString())

        if(explanation == null) {
            explanation = arrayOf("", "", "", "")
        }

        val unitname = view.findViewById<TextView>(R.id.unitexname)

        val lineid = intArrayOf(R.id.unitex0, R.id.unitex1, R.id.unitex2, R.id.unitex3)

        val explains = Array<TextView>(lineid.size) {
            view.findViewById(lineid[it])
        }

        explains[3].setPadding(0, 0, 0, StaticStore.dptopx(24f,requireActivity()))

        var name = MultiLangCont.get(u.forms[fid]) ?: u.forms[fid].name

        if (name == null)
            name = ""

        unitname.text = name

        for (i in explains.indices) {
            if (i >= explanation.size) {
                explains[i].visibility = View.GONE

                if(i > 1) {
                    explains[i - 1].setPadding(0, 0, 0, StaticStore.dptopx(24f, requireActivity()))
                }
            } else {
                if (explanation[i] != null)
                    explains[i].text = explanation[i]
            }
        }
        return view
    }
}